File: windows/Services/MsuPackApplier.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace LTTPRandomizerGenerator.Services;

public static class MsuPackApplier
{
    public static string? Apply(string romPath, Dictionary<string, string> tracks)
    {
        try
        {
            var baseName = Path.Combine(
                Path.GetDirectoryName(romPath)!,
                Path.GetFileNameWithoutExtension(romPath));

            // Write empty .msu marker file
            File.WriteAllBytes(baseName + ".msu", Array.Empty<byte>());

            // Copy each assigned PCM alongside the ROM
            foreach (var (slot, pcmPath) in tracks.OrderBy(t => int.Parse(t.Key)))
            {
                if (!File.Exists(pcmPath))
                    return $"PCM file for slot {slot} not found: {pcmPath}";

                File.Copy(pcmPath, $"{baseName}-{slot}.pcm", overwrite: true);
            }

            return null;
        }
        catch (Exception ex)
        {
            return $"MSU apply failed: {ex.Message}";
        }
    }
}

File: android/app/src/main/java/com/lttprandomizer/MsuPackApplier.kt
package com.lttprandomizer

import androidx.documentfile.provider.DocumentFile
import android.content.Context
import android.net.Uri
import java.io.File

object MsuPackApplier {
    fun apply(context: Context, outputDir: DocumentFile, romFileName: String, tracks: Map<String, String>): String? {
        return try {
            val baseName = romFileName.substringBeforeLast(".")

            // Write empty .msu marker file
            val msuFile = outputDir.createFile("application/octet-stream", "$baseName.msu")
                ?: return "Failed to create .msu marker file."
            // File is created empty by default via SAF

            // Copy each assigned PCM alongside the ROM
            for ((slot, pcmPath) in tracks.toSortedMap(compareBy { it.toInt() })) {
                val srcFile = File(pcmPath)
                if (!srcFile.exists()) return "PCM file for slot $slot not found: $pcmPath"

                val pcmName = "$baseName-$slot.pcm"
                // Remove existing file if present
                outputDir.findFile(pcmName)?.delete()

                val destFile = outputDir.createFile("application/octet-stream", pcmName)
                    ?: return "Failed to create PCM file: $pcmName"

                context.contentResolver.openOutputStream(destFile.uri)?.use { out ->
                    srcFile.inputStream().use { input -> input.copyTo(out) }
                } ?: return "Failed to write PCM file: $pcmName"
            }

            null
        } catch (ex: Exception) {
            "MSU apply failed: ${ex.message}"
        }
    }
}
